packages/kilt-did/src/types.chain.ts

```diff
@@ -0,0 +1,132 @@
+} from '@polkadot/types/interfaces'
+
+export interface DidVerificationKey extends Enum {
+  isEd25519: boolean
+  isSr25519: boolean
+  asEd25519: Vec<u8>
+  asSr25519: Vec<u8>
+  type: 'sr25519' | 'ed25519'
+  value: Vec<u8>
+}
+
+export interface DidEncryptionKey extends Enum {
+  isX25519: boolean
+  asX25519: Vec<u8>
+  type: 'x25519'
+  value: Vec<u8>
+}
+
+export interface VerificationKeyDetails extends Struct {
+  /// A verification key the DID controls.
+  verificationKey: DidVerificationKey
+  /// The block number in which the verification key was added to the DID.
+  blockNumber: BlockNumber
+}
+
+export type KeyId = Hash
+
+export type DidIdentifier = AccountId
+
+export interface UrlEncoding extends Struct {
+  payload: Vec<u8>
+}
+
+export interface Url extends Enum {
+  isHttp: boolean
+  isFtp: boolean
+  isIpfs: boolean
+  asHttp: UrlEncoding
+  asFtp: UrlEncoding
+  asIpfs: UrlEncoding
+}
+
+export interface DidPublicKey extends Enum {
+  isDidVerificationKey: boolean
+  asDidVerificationKey: DidVerificationKey
+  isDidEncryptionKey: boolean
+  asDidEncryptionKey: DidEncryptionKey
+  type: 'DidVerificationKey' | 'DidEncryptionKey'
+  value: DidVerificationKey | DidEncryptionKey
+}
+
+export interface DidPublicKeyDetails extends Struct {
+  key: DidPublicKey
+  blockNumber: BlockNumber
+}
+
+export interface DidDetails extends Struct {
+  authenticationKey: KeyId
+  keyAgreementKeys: BTreeSet<KeyId>
+  delegationKey: Option<KeyId>
+  attestationKey: Option<KeyId>
+  publicKeys: BTreeMap<KeyId, DidPublicKeyDetails>
+  endpointUrl: Option<Url>
+  lastTxCounter: u64
+}
+
+export interface DidCreationOperation extends Struct {
+  did: DidIdentifier
+  newAuthenticationKey: DidVerificationKey
+  newKeyAgreementKeys: BTreeSet<DidEncryptionKey>
+  newAttestationKey: Option<DidVerificationKey>
+  newDelegationKey: Option<DidVerificationKey>
+  newEndpointUrl: Option<Url>
+}
+
+export interface DidKeyUpdateAction extends Enum {
+  /// Do not change the verification key.
+  isIgnore: boolean
+  asIgnore: null
+  /// Change the verification key to the new one provided.
+  isChange: boolean
+  asChange: DidVerificationKey
+  /// Delete the verification key.
+  isDelete: boolean
+  asDelete: null
+}
+
+export interface DidUpdateOperation extends Struct {
+  did: DidIdentifier
+  newAuthenticationKey: Option<DidVerificationKey>
+  newKeyAgreementKeys: BTreeSet<DidEncryptionKey>
+  attestationKeyUpdate: DidKeyUpdateAction
+  delegationKeyUpdate: DidKeyUpdateAction
+  publicKeysToRemove: BTreeSet<KeyId>
+  newEndpointUrl: Option<Url>
+  txCounter: u64
+}
+
+export interface DidDeletionOperation extends Struct {
+  did: DidIdentifier
+  txCounter: u64
+}
+
+export interface DidSignature extends Enum {
+  isEd25519: boolean
+  isSr25519: boolean
+  asEd25519: Vec<u8>
+  asSr25519: Vec<u8>
+}
+
+export interface DidAuthorizedCallOperation extends Struct {
+  did: DidIdentifier
+  txCounter: u64
+  call: DidCallable
+}
+
+type DidCallable = Call
```

Comment:
Maybe let's keep all the type definitions together, and the top of the file sounds like a good place to do that.

Reply:
I'm not sure what this refers to ? But I should probably comment on the longer term plans for these types files. The `types.ts` file is supposed to be moved to the types package (I could also do that now), while the `types.chain.ts` is tightly linked to the types-definitions repo and may live there in the future, as this makes it easier to maintain typescript interfaces along with the substrate types they represent. Additionally, there is some (currently broken) tooling to convert the substrate types into interfaces.

Reply:
I just mean that we should group together all the `type` definitions within the same file, where possible 😄

Reply:
oh, yeah, we can do that. doesn't make a difference to me